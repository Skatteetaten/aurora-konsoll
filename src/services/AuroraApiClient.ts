import { default as ApolloClient } from 'apollo-boost';

import { tokenStore } from './TokenStore';

import {
  APPLICATIONS_QUERY,
  IApplications,
  IPodResource,
  IUserAffiliationsQuery,
  USER_AFFILIATIONS_QUERY
} from './auroraApiClient/queries';

export interface IAuroraApiClient {
  findUserAndAffiliations: () => Promise<IUserAffiliationResult>;
  findAllApplicationsForAffiliations: (
    affiliations: string[]
  ) => Promise<IApplicationResult[]>;
}

export default class AuroraApiClient implements IAuroraApiClient {
  private client: ApolloClient<{}>;

  constructor(graphQLUrl: string) {
    const client = new ApolloClient({
      request: async operations => {
        const token = tokenStore.getToken();
        operations.setContext({
          headers: {
            Authorization: token ? `Bearer ${token}` : ''
          }
        });
      },
      uri: graphQLUrl
    });
    client.defaultOptions = {
      query: {
        fetchPolicy: 'network-only'
      }
    };

    this.client = client;
  }

  public async findUserAndAffiliations(): Promise<IUserAffiliationResult> {
    const result = await this.client.query<IUserAffiliationsQuery>({
      query: USER_AFFILIATIONS_QUERY
    });

    return {
      affiliations: result.data.affiliations.edges.map(edge => edge.node.name),
      user: result.data.currentUser.name
    };
  }

  public async findAllApplicationsForAffiliations(
    affiliations: string[]
  ): Promise<IApplicationResult[]> {
    const result = await this.client.query<IApplications>({
      query: APPLICATIONS_QUERY,
      variables: {
        affiliations
      }
    });

    return result.data.applications.edges.reduce((acc, { node }) => {
      const { applicationInstances } = node;
      const apps = applicationInstances.map(app => ({
        affiliation: app.affiliation.name,
        environment: app.environment,
        name: node.name,
        pods: app.details.podResources,
        statusCode: app.status.code,
        version: {
          auroraVersion: app.version.auroraVersion,
          deployTag: app.version.deployTag
        }
      }));
      return [...acc, ...apps];
    }, []);
  }
}

export interface IUserAffiliationResult {
  affiliations: string[];
  user: string;
}

export interface IApplicationResult {
  affiliation: string;
  name: string;
  environment: string;
  statusCode: string;
  version: {
    auroraVersion: string;
    deployTag: string;
  };
  pods: IPodResource[];
}
