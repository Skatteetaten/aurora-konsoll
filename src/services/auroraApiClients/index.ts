export * from './applicationDeploymentClient/client';
export * from './imageRepositoryClient/client';
export * from './netdebugClient/client';
export * from './userSettingsClient/client';
export * from './databaseClient/client';
export * from './websealClient/client';
export * from './certificateClient/client';
