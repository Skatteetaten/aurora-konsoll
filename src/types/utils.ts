// Taken from http://www.typescriptlang.org/docs/handbook/advanced-types.html
export type Omit<T, K> = Pick<T, Exclude<keyof T, K>>;