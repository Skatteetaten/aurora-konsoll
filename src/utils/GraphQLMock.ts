import cors from 'cors';
import express from 'express';
import { Server } from 'http';
import GoboClient from 'services/GoboClient';

type ResponseMock = any;

interface IGraphQLResponseMock {
  [queryName: string]: ResponseMock;
}

export function goboClientMock(url: string): GoboClient {
  return new GoboClient({
    url
  });
}

export class GraphQLSeverMock {
  public port: number;
  public graphQLUrl: string;
  private server: Server;
  private responses: IGraphQLResponseMock[];

  constructor(port?: number) {
    const PORT = port || randomPort(40000, 50000);
    const ENDPOINT = '/graphql';

    const server = express();
    server.use(cors());
    server.use(express.json());

    server.post(ENDPOINT, (req, res) => {
      const { operationName } = req.body;
      const responseMock = this.responses[operationName];

      if (!responseMock) {
        res.sendStatus(404);
      } else {
        res.send(responseMock);
      }
    });

    this.responses = [];
    this.server = server.listen(PORT);
    this.port = PORT;
    this.graphQLUrl = `http://localhost:${PORT}${ENDPOINT}`;
  }

  public close() {
    this.server.close();
  }

  public putResponse(queryName: string, response: any) {
    this.responses[queryName] = response;
  }
}

function randomPort(start: number, stop: number): number {
  const random = Math.random() * (stop - start) + start;
  return Math.trunc(random);
}
