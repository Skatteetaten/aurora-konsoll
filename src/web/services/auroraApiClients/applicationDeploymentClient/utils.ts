import YAML from 'yaml';

export function changeVersionInFile(
  fileName: string,
  fileContent: string,
  version: string
) {
  if (fileName.endsWith('json')) {
    const parsedApplicationFile: JSON = JSON.parse(fileContent);
    return JSON.stringify(
      { ...parsedApplicationFile, version: version },
      null,
      2
    );
  }
  try {
    const parsedApplicationFile = YAML.parseDocument(fileContent);
    parsedApplicationFile.set('version', version);
    return parsedApplicationFile.toString();
  } catch {
    return undefined;
  }
}
