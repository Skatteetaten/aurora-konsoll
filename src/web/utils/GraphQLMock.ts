import cors from 'cors';
import express from 'express';
import { Server } from 'http';
import GoboClient from 'web/services/GoboClient';
import { IApiClients } from 'web/models/AuroraApi';
import {
  ApplicationDeploymentClient,
  CertificateClient,
  DatabaseClient,
  ImageRepositoryClient,
  NetdebugClient,
  UserSettingsClient,
  WebsealClient,
} from 'web/services/auroraApiClients';
import { Socket } from 'net';
import { DnsClient } from 'web/services/auroraApiClients/dnsClient/client';

type ResponseMock = any;

interface IGraphQLResponseMock {
  [queryName: string]: ResponseMock;
}

export class GraphQLSeverMock {
  public port: number;
  public graphQLUrl: string;
  private server: Server;
  private responses: IGraphQLResponseMock[];
  private serverSockets = new Set<Socket>();

  constructor(port?: number) {
    const PORT = port || randomPort(40000, 50000);
    const ENDPOINT = '/graphql';

    const server = express();
    server.use(cors());
    server.use(express.json());

    server.post(ENDPOINT, (req, res) => {
      const { operationName } = req.body;
      const responseMock = this.responses[operationName];
      if (!responseMock) {
        res.sendStatus(404);
      } else {
        res.send(responseMock);
      }
    });

    this.responses = [];
    this.server = server.listen(PORT);
    this.port = PORT;
    this.graphQLUrl = `http://localhost:${PORT}${ENDPOINT}`;

    // Keeps track of open connections that must be closed after test is done.
    this.server.on('connection', (socket) => {
      this.serverSockets.add(socket);
      socket.on('close', () => {
        this.serverSockets.delete(socket);
      });
    });
  }

  /**
   * Closes all connections and the server.
   */
  public close(done: jest.DoneCallback) {
    this.server.close(() => {
      done();
    });
    for (const socket of this.serverSockets.values()) {
      socket.destroy();
    }
  }

  public putResponse(queryName: string, response: any) {
    this.responses[queryName] = response;
  }
}

function randomPort(start: number, stop: number): number {
  const random = Math.random() * (stop - start) + start;
  return Math.trunc(random);
}

export function goboClientMock(url: string): GoboClient {
  return new GoboClient({ url });
}

export function getApiClientsMock(serverMock: GraphQLSeverMock): IApiClients {
  const clientMock = goboClientMock(serverMock.graphQLUrl);
  return {
    applicationDeploymentClient: new ApplicationDeploymentClient(clientMock),
    certificateClient: new CertificateClient(clientMock),
    databaseClient: new DatabaseClient(clientMock),
    imageRepositoryClient: new ImageRepositoryClient(clientMock),
    netdebugClient: new NetdebugClient(clientMock),
    userSettingsClient: new UserSettingsClient(clientMock),
    websealClient: new WebsealClient(clientMock),
    dnsClient: new DnsClient(clientMock),
  };
}
