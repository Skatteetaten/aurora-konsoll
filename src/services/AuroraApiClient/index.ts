import ApolloClient from 'apollo-boost';

import { tokenStore } from '../TokenStore';

import {
  IUserAffiliationsQuery,
  USER_AFFILIATIONS_QUERY
} from './queries/user-affiliations-query';

import {
  APPLICATIONS_QUERY,
  IApplications
} from './queries/applications-query';
import { ITagsQuery, TAGS_QUERY } from './queries/tag-query';
import { IApplication, ITagsPaged, IUserAffiliationResult } from './types';
import { IAuroraApiClient } from './types';

export default class AuroraApiClient implements IAuroraApiClient {
  private client: ApolloClient<{}>;

  constructor(graphQLUrl: string) {
    const client = new ApolloClient({
      request: async operations => {
        const token = tokenStore.getToken();
        operations.setContext({
          headers: {
            Authorization: token ? `Bearer ${token}` : ''
          }
        });
      },
      uri: graphQLUrl
    });
    client.defaultOptions = {
      query: {
        fetchPolicy: 'network-only'
      }
    };

    this.client = client;
  }

  public async findTagsPaged(
    repository: string,
    cursor?: string
  ): Promise<ITagsPaged> {
    const result = await this.client.query<ITagsQuery>({
      query: TAGS_QUERY,
      variables: {
        cursor,
        repositories: [repository]
      }
    });

    const { imageRepositories } = result.data;

    if (!(imageRepositories && imageRepositories.length > 0)) {
      throw new Error(`Could not find tags for repository ${repository}`);
    }

    const [mainRepo] = imageRepositories;
    const { edges, pageInfo } = mainRepo.tags;

    return {
      endCursor: pageInfo.endCursor,
      hasNextPage: pageInfo.hasNextPage,
      tags: edges.map(edge => ({
        lastModified: edge.node.lastModified,
        name: edge.node.name
      }))
    };
  }

  public async findUserAndAffiliations(): Promise<IUserAffiliationResult> {
    const result = await this.client.query<IUserAffiliationsQuery>({
      query: USER_AFFILIATIONS_QUERY
    });

    return {
      affiliations: result.data.affiliations.edges.map(edge => edge.node.name),
      user: result.data.currentUser.name
    };
  }

  public async findAllApplicationsForAffiliations(
    affiliations: string[]
  ): Promise<IApplication[]> {
    const result = await this.client.query<IApplications>({
      query: APPLICATIONS_QUERY,
      variables: {
        affiliations
      }
    });

    return result.data.applications.edges.reduce((acc, { node }) => {
      const { applicationInstances, imageRepository } = node;
      const apps = applicationInstances.map(app => ({
        affiliation: app.affiliation.name,
        environment: app.environment,
        name: node.name,
        pods: app.details.podResources,
        repository: imageRepository.repository,
        statusCode: app.status.code,
        version: {
          auroraVersion: app.version.auroraVersion,
          deployTag: app.version.deployTag
        }
      }));
      return [...acc, ...apps];
    }, []);
  }
}
