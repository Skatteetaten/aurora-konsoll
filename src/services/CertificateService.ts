import {
  ICertificateResult,
  ICertificateView,
  IDetailsListContent
} from 'models/certificates';
import { createDate, getLocalDate, isDate } from 'utils/date';
import { SortDirection } from './DatabaseSchemaService';

export const certificateColumns = (): IDetailsListContent[] => [
  {
    fieldName: 'id',
    isResizable: true,
    key: 0,
    maxWidth: 150,
    minWidth: 100,
    name: 'ID',
    iconName: ''
  },
  {
    fieldName: 'dn',
    isResizable: true,
    key: 1,
    maxWidth: 500,
    minWidth: 400,
    name: 'DN',
    iconName: ''
  },
  {
    fieldName: 'revokedDate',
    isResizable: true,
    key: 2,
    maxWidth: 300,
    minWidth: 150,
    name: 'Revoked',
    iconName: ''
  },
  {
    fieldName: 'issuedDate',
    isResizable: true,
    key: 3,
    maxWidth: 300,
    minWidth: 150,
    name: 'Opprettet',
    iconName: ''
  },
  {
    fieldName: 'expiresDate',
    isResizable: true,
    key: 4,
    maxWidth: 300,
    minWidth: 150,
    name: 'UtlÃ¸per',
    iconName: ''
  }
];

export const filterCertificateView = (filter: string) => {
  return (v: ICertificateView): boolean =>
    v.dn.includes(filter) ||
    v.id.toString().includes(filter) ||
    v.issuedDate.includes(filter) ||
    (!v.revokedDate || v.revokedDate === null
      ? false
      : v.revokedDate.includes(filter)) ||
    (!v.expiresDate || v.expiresDate === null
      ? false
      : v.expiresDate.includes(filter));
};

export const defaultSortDirections: SortDirection[] = new Array<SortDirection>(
  certificateColumns().length
).fill(SortDirection.NONE);

export default class CertificateService {
  public sortNextAscending(sortDirection: SortDirection): boolean {
    return (
      sortDirection === SortDirection.NONE ||
      sortDirection === SortDirection.DESC
    );
  }

  public createColumns(
    index: number,
    sortDirection: SortDirection
  ): IDetailsListContent[] {
    const columns = certificateColumns();
    if (index > -1) {
      const currentCol = columns[index];
      if (
        sortDirection === SortDirection.NONE ||
        sortDirection === SortDirection.DESC
      ) {
        currentCol.iconName = 'Down';
      } else if (sortDirection === SortDirection.ASC) {
        currentCol.iconName = 'Up';
      }
    }
    return columns;
  }

  public sortItems(
    viewItems: ICertificateView[],
    prevSortDirection: SortDirection,
    name: string | number | symbol
  ): ICertificateView[] {
    return viewItems.slice(0).sort((a: any, b: any) => {
      const valueA = this.lowerCaseIfString(a[name]);
      const valueB = this.lowerCaseIfString(b[name]);
      if (valueA === valueB) {
        return 0;
      } else if (isDate(valueA) || isDate(valueB)) {
        const dateA = createDate(valueA).getTime();
        const dateB = createDate(valueB).getTime();
        return this.sortNextAscending(prevSortDirection)
          ? dateB - dateA
          : dateA - dateB;
      } else {
        return (this.sortNextAscending(prevSortDirection)
        ? valueA < valueB
        : valueA > valueB)
          ? 1
          : -1;
      }
    });
  }

  public updatedItems = (data: ICertificateResult): ICertificateView[] =>
    data.certificates.map(
      (it): ICertificateView => {
        return {
          id: it.id,
          dn: it.dn,
          issuedDate: getLocalDate(it.issuedDate),
          revokedDate: !!it.revokedDate ? getLocalDate(it.revokedDate) : '-',
          expiresDate: !!it.expiresDate ? getLocalDate(it.expiresDate) : '-'
        };
      }
    );

  public filteredItems = (
    filter: string,
    viewItems: ICertificateView[]
  ): ICertificateView[] => {
    return viewItems.filter(filterCertificateView(filter));
  };

  private lowerCaseIfString(value: any): any {
    return typeof value === 'string' ? (value as string).toLowerCase() : value;
  }
}
