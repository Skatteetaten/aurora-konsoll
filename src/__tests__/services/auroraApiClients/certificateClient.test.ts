import { CertificateClient } from 'services/auroraApiClients';
import { goboClientMock, GraphQLSeverMock } from 'utils/GraphQLMock';

import * as getCertificates from './__responses__/certificateClient/getCertificates.json';

const serverMock = new GraphQLSeverMock();
const clientMock = goboClientMock(serverMock.graphQLUrl);
const certificateClient = new CertificateClient(clientMock);

afterAll((done) => {
  serverMock.close(done);
});

describe('getCertificates', () => {
  it('should fetch certificates from GraphQL server', async () => {
    serverMock.putResponse('getCertificates', getCertificates);

    const result = await certificateClient.getCertificates();
    expect(result).toMatchSnapshot();
  });
});
