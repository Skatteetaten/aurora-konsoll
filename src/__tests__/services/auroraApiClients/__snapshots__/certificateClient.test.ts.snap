// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCertificates should fetch certificates from GraphQL server 1`] = `
Object {
  "certificates": Array [
    Object {
      "dn": "Test1",
      "expiresDate": "2019-03-27T10:58:25.188860Z",
      "id": "123",
      "issuedDate": "2019-03-27T10:58:25.188859Z",
      "revokedDate": "2019-03-27T10:58:25.188860Z",
    },
    Object {
      "dn": "Test2",
      "expiresDate": "2019-03-27T10:58:25.188865Z",
      "id": "234",
      "issuedDate": "2019-03-27T10:58:25.188865Z",
      "revokedDate": "2019-03-27T10:58:25.188865Z",
    },
  ],
  "totalCount": 2,
}
`;
